crypto/chacha/asm/chacha-riscv64-zbb-zvkb.pl

```diff
@@ -163,139 +227,232 @@ sub chacha_quad_round_group {
     @{[vxor_vv $V20, $V20, $V4]}
+    lw $T0, 16($KEY)
     @{[vxor_vv $V21, $V21, $V5]}
+    lw $T1, 20($KEY)
     @{[vxor_vv $V22, $V22, $V6]}
+    lw $T2, 24($KEY)
     @{[vxor_vv $V23, $V23, $V7]}
 
     # save the top-half of output
-    @{[vssseg_nf_e32_v 8, $V16, $OUTPUT, $STRIDE]}
+    li $T3, 64
+    @{[vssseg_nf_e32_v 8, $V16, $OUTPUT, $T3]}
 
     # add chacha bottom-half initial block states
-    @{[vadd_vx $V8, $V8, $KEY4]}
-    @{[vadd_vx $V9, $V9, $KEY5]}
-    @{[vadd_vx $V10, $V10, $KEY6]}
-    @{[vadd_vx $V11, $V11, $KEY7]}
+    @{[vadd_vx $V8, $V8, $T0]}
+    add $STATE8, $STATE8, $T0
+    lw $T3, 28($KEY)
+    @{[vadd_vx $V9, $V9, $T1]}
+    add $STATE9, $STATE9, $T1
+    lw $T0, 4($COUNTER)
+    @{[vadd_vx $V10, $V10, $T2]}
+    add $STATE10, $STATE10, $T2
+    lw $T1, 8($COUNTER)
+    @{[vadd_vx $V11, $V11, $T3]}
+    add $STATE11, $STATE11, $T3
+    lw $T2, 12($COUNTER)
     @{[vid_v $V0]}
-    @{[vadd_vx $V12, $V12, $COUNTER0]}
-    @{[vadd_vx $V13, $V13, $COUNTER1]}
-    @{[vadd_vx $V14, $V14, $NONCE0]}
-    @{[vadd_vx $V15, $V15, $NONCE1]}
+    add $STATE12, $STATE12, $CURRENT_COUNTER
+    @{[vadd_vx $V12, $V12, $CURRENT_COUNTER]}
+    add $STATE12, $STATE12, $VL
+    @{[vadd_vx $V13, $V13, $T0]}
+    add $STATE13, $STATE13, $T0
+    @{[vadd_vx $V14, $V14, $T1]}
+    add $STATE14, $STATE14, $T1
+    @{[vadd_vx $V15, $V15, $T2]}
+    add $STATE15, $STATE15, $T2
     @{[vadd_vv $V12, $V12, $V0]}
     # xor with the bottom-half input
     @{[vxor_vv $V24, $V24, $V8]}
+    sw $STATE8, 32(sp)
     @{[vxor_vv $V25, $V25, $V9]}
+    sw $STATE9, 36(sp)
     @{[vxor_vv $V26, $V26, $V10]}
+    sw $STATE10, 40(sp)
     @{[vxor_vv $V27, $V27, $V11]}
+    sw $STATE11, 44(sp)
     @{[vxor_vv $V29, $V29, $V13]}
+    sw $STATE12, 48(sp)
     @{[vxor_vv $V28, $V28, $V12]}
+    sw $STATE13, 52(sp)
     @{[vxor_vv $V30, $V30, $V14]}
+    sw $STATE14, 56(sp)
     @{[vxor_vv $V31, $V31, $V15]}
+    sw $STATE15, 60(sp)
 
     # save the bottom-half of output
-    addi $T0, $OUTPUT, 32
-    @{[vssseg_nf_e32_v 8, $V24, $T0, $STRIDE]}
+    li $T0, 64
+    addi $T1, $OUTPUT, 32
+    @{[vssseg_nf_e32_v 8, $V24, $T1, $T0]}
 
-    # update counter
-    add $COUNTER0, $COUNTER0, $VL
-    sub $LEN, $LEN, $VL
-    # increase offset for `4 * 16 * VL = 64 * VL`
+    # the computed vector parts: `64 * VL`
     slli $T0, $VL, 6
+
     add $INPUT, $INPUT, $T0
     add $OUTPUT, $OUTPUT, $T0
+    sub $LEN, $LEN, $T0
+    add $CURRENT_COUNTER, $CURRENT_COUNTER, $VL
+
+    # process the scalar data block
+    addi $CURRENT_COUNTER, $CURRENT_COUNTER, 1
+    li $T0, 64
+    @{[minu $T1, $LEN, $T0]}
+    sub $LEN, $LEN, $T1
+    mv $T2, sp
+.Lscalar_data_loop:
+    @{[vsetvli $VL, $T1, "e8", "m8", "ta", "ma"]}
+    @{[vle8_v $V8, $INPUT]}
+    @{[vle8_v $V16, $T2]}
+    @{[vxor_vv $V8, $V8, $V16]}
+    @{[vse8_v $V8, $OUTPUT]}
+    add $INPUT, $INPUT, $VL
+    add $OUTPUT, $OUTPUT, $VL
+    add $T2, $T2, $VL
+    sub $T1, $T1, $VL
+    bnez $T1, .Lscalar_data_loop
```

Comment:
A question though, since VLEN>=128 is assumed then `vsetvli` will make VLMAX>=128, and from `minu` we have T1<=64, so `VL==T1` and this `bnez` will never branch.

Reply:
I test with qemu using vlen=32, 64 and other larger vlen settings before. That's why here is a branch here.
We could omit the branch here, but it's just a trivial one.